### PR TITLE
add position to ruler prop

### PR DESCRIPTION
**Pull request by jasmineleeson**
- add position to ruler prop so that `InteractionLayer` can set the initial position of the ruler based on the timestamp provided
- add `onFetchData` callback to `DataProvider`
- add `onMouseOut` and `onBlur` callbacks to `LineChart`

### CODE REVIEW

File: src/components/DataProvider/index.js [hunk 1]
```diff
@@ -602,6 +604,7 @@ DataProvider.propTypes = {
   // xSubDomain => newXSubDomain
   // function to allow limitation of the value of xSubDomain
   limitXSubDomain: PropTypes.func,
+  onFetchData: PropTypes.func,
```
Comment (evancharlton) [inline src/components/DataProvider/index.js]: What does this signature look like? What does it do? Why would I use it? (please jsdoc this)

File: src/utils/proptypes.js [hunk 2]
```diff
@@ -60,6 +60,8 @@ export const rulerPropType = PropTypes.shape({
   visible: PropTypes.bool,
   xLabel: PropTypes.func.isRequired,
   yLabel: PropTypes.func.isRequired,
+  // a timestamp representing the initial position of the ruler
+  position: PropTypes.number,
```
Comment (evancharlton) [inline src/utils/proptypes.js]: `position` is super vague -- what about `timestamp` or something?

Comment (evancharlton) [reply src/utils/proptypes.js]: (and while I'm looking at this, I see that `xLabel` should be `timeLabel` . . .)

File: stories/LineChart.stories.js [hunk 3]
```diff
@@ -735,6 +735,7 @@ storiesOf('LineChart', module)
             `${point.name}: ${Number.parseFloat(point.value).toFixed(3)}`,
           xLabel: point =>
             moment(point.timestamp).format('DD-MM-YYYY HH:mm:ss'),
+          position: Date.now() - 1000 * 10,
```
Comment (evancharlton) [inline stories/LineChart.stories.js]: Could you add this as a separate story? Or maybe just a second `<LineChart>` here. I'd like to be able to compare the behavior with and without this prop set.

File: src/components/LineChart/index.js [hunk 4]
```diff
@@ -94,6 +96,8 @@ const defaultProps = {
   },
   crosshair: true,
   onMouseMove: null,
+  onMouseOut: null,
```
Comment (evancharlton) [inline src/components/LineChart/index.js]: Can you add a story for these? They just need to call `action('onMouseOut')` for example.

File: src/components/InteractionLayer/index.js [hunk 5]
```diff
@@ -462,24 +448,76 @@ class InteractionLayer extends React.Component {
...
+    const output = this.getDataForCoordinate(xpos, ypos, true);
+    this.setState({
+      area: {
+        ...area,
+        end: output,
+      },
+    });
+  };
+
+  processExternalMouseMove = (xpos, ypos) => {
```
Comment (evancharlton) [inline src/components/InteractionLayer/index.js]: How about `processActualMouseMove` or something? I feel like the "external" keyword is becoming overloaded.

File: src/components/InteractionLayer/index.js [hunk 6]
```diff
@@ -462,24 +448,76 @@ class InteractionLayer extends React.Component {
...
+    this.processMouseMove(xpos, ypos, true);
+  };
 
+  processMouseMove = (xpos, ypos, external = false) => {
```
Comment (evancharlton) [inline src/components/InteractionLayer/index.js]: oh boy, `external: boolean` seems _dangerous_. Why not pass the actual event in?

**Comment by cognite-cicd** [comment]
[pr-server]
The storybook for this PR is hosted on https://griff-react-pr-195.surge.sh

File: stories/LineChart.stories.js [hunk 7]
```diff
@@ -875,4 +902,23 @@ storiesOf('LineChart', module)
...
+      series={[{ id: 1, color: 'steelblue' }, { id: 2, color: 'maroon' }]}
+      xDomain={staticXDomain}
+      xSubDomain={[
+        Date.now() - 1000 * 60 * 60 * 24 * 30,
+        Date.now() - 1000 * 60 * 60 * 24 * 10,
+      ]}
+    >
+      <LineChart
+        height={CHART_HEIGHT}
+        onMouseOut={e => {
```
Comment (evancharlton) [inline stories/LineChart.stories.js]: nit: you can just do `onMouseOut={action('onMouseOut')}` since `action` returns a function

**Comment by cognite-cicd** [comment]
[pr-server]
The storybook for this PR is hosted on https://griff-react-pr-195.surge.sh

**Comment by cognite-cicd** [comment]
[pr-server]
The storybook for this PR is hosted on https://griff-react-pr-195.surge.sh

File: src/components/DataProvider/index.js [hunk 8]
```diff
@@ -453,7 +453,9 @@ export default class DataProvider extends Component {
         [id]: { ...loaderConfig },
       };
     }
-    this.setState(stateUpdates);
+    this.setState(stateUpdates, () => {
+      this.props.onFetchData();
```
Comment (valerii-cognite) [inline src/components/DataProvider/index.js]: Would be cleaner, if `onFetchData` will be null as a default and will be called only if has been passed down

File: src/components/InteractionLayer/index.js [hunk 9]
```diff
@@ -419,24 +425,58 @@ class InteractionLayer extends React.Component {
...
+    });
+  };
 
+  processMouseMove = (xpos, ypos, e = null) => {
+    const rulerPoints = this.setRulerPoints(xpos);
+    this.setArea(xpos, ypos);
+    const { onMouseMove } = this.props;
     if (onMouseMove) {
-      onMouseMove({ points: newPoints, xpos, ypos });
+      onMouseMove({ points: rulerPoints, xpos, ypos, e });
```
Comment (valerii-cognite) [inline src/components/InteractionLayer/index.js]: Talking of API, I would expect `event` to be a first parameter, and then an object of others

File: src/components/InteractionLayer/index.js [hunk 10]
```diff
@@ -75,8 +75,9 @@ class InteractionLayer extends React.Component {
     series: [],
     ruler: {
       visible: false,
-      xLabel: () => {},
+      timeLabel: () => {},
       yLabel: () => {},
+      timestamp: null,
```
Comment (valerii-cognite) [inline src/components/InteractionLayer/index.js]: I would imagine at some point we will apply ruler for `ScatterPlot`, so this prop will probably be revisited.

**Pull request merged**